File: rocket-chat-core/src/main/java/chat/rocket/core/interactors/RoomInteractor.java
package chat.rocket.core.interactors;

import io.reactivex.Flowable;

import java.util.List;
import chat.rocket.core.models.Room;
import chat.rocket.core.repositories.RoomRepository;

public class RoomInteractor {

  private final RoomRepository roomRepository;

  public RoomInteractor(RoomRepository roomRepository) {
    this.roomRepository = roomRepository;
  }

  public Flowable<Integer> getTotalUnreadMentionsCount() {
    return roomRepository.getAll()
        .flatMap(rooms -> Flowable.fromIterable(rooms)
            .filter(room -> room.isOpen() && room.isAlert())
            .map(Room::getUnread)
            .defaultIfEmpty(0)
            .reduce((unreadCount, unreadCount2) -> unreadCount + unreadCount2)
            .toFlowable());
  }

  public Flowable<Long> getTotalUnreadRoomsCount() {
    return roomRepository.getAll()
        .flatMap(rooms -> Flowable.fromIterable(rooms)
            .filter(room -> room.isOpen() && room.isAlert())
            .count()
            .toFlowable());
  }

  public Flowable<List<Room>> getOpenRooms() {
    return roomRepository.getAll()
        .flatMap(rooms -> Flowable.fromIterable(rooms)
            .filter(Room::isOpen)
            .toList()
            .toFlowable());
  }
}

File: rocket-chat-core/src/main/java/chat/rocket/core/interactors/RoomInteractor.kt
package chat.rocket.core.interactors

import io.reactivex.Flowable
import chat.rocket.core.SortDirection
import chat.rocket.core.models.Room
import chat.rocket.core.repositories.RoomRepository

class RoomInteractor(private val roomRepository: RoomRepository) {

    fun getTotalUnreadMentionsCount(): Flowable<Int> {
        return roomRepository.all
                .flatMap { rooms ->
                    Flowable.fromIterable(rooms)
                            .filter { room -> room.isOpen && room.isAlert }
                            .map { it.unread }
                            .defaultIfEmpty(0)
                            .reduce { unreadCount, unreadCount2 -> unreadCount + unreadCount2 }
                            .toFlowable()
                }
    }

    fun getTotalUnreadRoomsCount(): Flowable<Long> {
        return roomRepository.all
                .flatMap { rooms ->
                    Flowable.fromIterable(rooms)
                            .filter { room -> room.isOpen && room.isAlert }
                            .count()
                            .toFlowable()
                }
    }

    fun getOpenRooms(): Flowable<List<Room>> {
        return roomRepository.all
                .flatMap { rooms ->
                    Flowable.fromIterable(rooms)
                            .filter { it.isOpen }
                            .toList()
                            .toFlowable()
                }
    }

    fun getRoomsWithNameLike(name: String): Flowable<List<Room>> {
        return roomRepository.getSortedLikeName(name, SortDirection.DESC, 5)
    }
}
